plutus-tx/src/PlutusTx/Blueprint/Schema.hs

```diff
@@ -0,0 +1,417 @@
+          , Just $ "values" .= valueSchema
+          , fmap ("minItems" .=) minItems
+          , fmap ("maxItems" .=) maxItems
+          , fmap ("title" .=) title
+          , fmap ("description" .=) description
+          , fmap ("$comment" .=) comment
+          ]
+    SchemaConstructor title description comment index fieldSchemas ->
+      Aeson.object
+        $ catMaybes
+          [ Just ("dataType" .= ("constructor" :: String))
+          , Just $ "constructor" .= index
+          , case fieldSchemas of
+              [] -> Nothing
+              xs -> Just $ "fields" .= xs
+          , fmap ("title" .=) title
+          , fmap ("description" .=) description
+          , fmap ("$comment" .=) comment
+          ]
+    SchemaBuiltInData title description comment ->
+      Aeson.object
+        $ catMaybes
+          [ fmap ("title" .=) title
+          , fmap ("description" .=) description
+          , fmap ("$comment" .=) comment
+          ]
+    SchemaBuiltInUnit title description comment ->
+      dataType title description comment "#unit"
+    SchemaBuiltInBoolean title description comment ->
+      dataType title description comment "#boolean"
+    SchemaBuiltInInteger title description comment ->
+      dataType title description comment "#integer"
+    SchemaBuiltInBytes title description comment ->
+      dataType title description comment "#bytes"
+    SchemaBuiltInString title description comment ->
+      dataType title description comment "#string"
+    SchemaBuiltInPair title description comment schema1 schema2 ->
+      Aeson.object
+        $ catMaybes
+          [ Just $ "dataType" .= ("#pair" :: String)
+          , Just $ "fst" .= schema1
+          , Just $ "snd" .= schema2
+          , fmap ("title" .=) title
+          , fmap ("description" .=) description
+          , fmap ("$comment" .=) comment
+          ]
+    SchemaBuiltInList title description comment schema ->
+      Aeson.object
+        $ catMaybes
+          [ Just $ "dataType" .= ("#list" :: String)
+          , Just $ "items" .= schema
+          , fmap ("title" .=) title
+          , fmap ("description" .=) description
+          , fmap ("$comment" .=) comment
+          ]
+    SchemaOneOf schemas -> Aeson.object ["oneOf" .= schemas]
+    SchemaAnyOf schemas -> Aeson.object ["anyOf" .= schemas]
+    SchemaAllOf schemas -> Aeson.object ["allOf" .= schemas]
+    SchemaNot schema -> Aeson.object ["not" .= schema]
+    SchemaRef ref -> Aeson.object ["$ref" .= ("#/definitions/" ++ ref)]
+   where
+    dataType :: Maybe String -> Maybe String -> Maybe String -> String -> Aeson.Value
+    dataType title description comment ty =
+      Aeson.object
+        $ catMaybes
+          [ Just $ "dataType" .= ty
+          , fmap ("title" .=) title
+          , fmap ("description" .=) description
+          , fmap ("$comment" .=) comment
+          ]
+
+type HasSchemaDefinition :: forall a. a -> [a] -> Constraint
+type family HasSchemaDefinition n xs where
```

Comment:
This is another thing that could live in a different PR. Without seeing how it is used, it's difficult to tell whether this is worth having.

Is the idea to first extract all the definitions into a type-level list, before beginning to generate the blueprint? An alternative would be to use a writer monad that collects the `[Definition]` as you go, then populating the definition section with the `[Definition]` collected. It's unclear whether this is better.